dist/tools/mkconstfs/mkconstfs2.py

```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env python3
+
+import os
+import sys
+import argparse
+import io
+import itertools
+import mmap
+import shutil
+from binascii import hexlify
+
+C_HEADER = """/* This file was automatically generated by mkconstfs.
+ * !!!! DO NOT EDIT !!!!!
+ */
+
+#include <stdint.h>
+#include "fs/constfs.h"
+
+"""
+
+FILE_TEMPLATE = """ {{
+  .path = "{target_name}",
+  .data = {mangled_name},
+  .size = sizeof({mangled_name})
+ }},
+"""
+
+C_FOOTER = """
+static const constfs_t _fs_data = {{
+    .files = _files,
+    .nfiles = sizeof(_files) / sizeof(_files[0]),
+}};
+
+vfs_mount_t {constfs_name} = {{
+    .fs = &constfs_file_system,
+    .mount_point = "{mount_pount}",
+    .private_data = (void *)&_fs_data,
+}};
+"""
+
+FILES_DECL = """
+static const constfs_file_t _files[] = {
+"""
+
+BLOB_DECL = """
+/** {fname} **/
+static const uint8_t {varname}[] = {{
+"""
+
+
+def mkconstfs(files, root_path, mount_point, constfs_name):
+    """Generate a C file containing a constant file system
+
+    Return
+    ------
+
+    chunks: Iterator yielding fragments of the of the output file.
+    """
+
+    filemap = {f: os.path.relpath(f, root_path) for f in files}
+
+    yield C_HEADER
+    yield from itertools.chain.from_iterable(
+                print_file_data(*fl) for fl in filemap.items())
+
+    yield FILES_DECL
+
+    yield from (FILE_TEMPLATE.format(target_name=relp, mangled_name=mangle_name(relp))
+                for relp in sorted(filemap.values()))
+
+    yield "};\n"
+
+    yield C_FOOTER.format(constfs_name=constfs_name, mount_pount=mount_point)
+
+
+def mangle_name(fname):
+    return "_"+"".join(c if c.isalnum() else "_{:04x}".format(ord(c))
+                       for c in fname)
+
+
+def print_file_data(local_fname, target_fname):
+    """Convert a file into a static C array:
+
+    Parameters
+    ----------
+
+    local_fname: real file name (where the file is on this machine's fs)
+    target_fname: name that the file will have in the constfs.
+    output_file: File-like object where the array will be written.
+
+    Return
+    ------
+
+    chunks: Iterator yielding fragments of the of the output text.
+    """
+
+    varname = mangle_name(target_fname)
+    yield BLOB_DECL.format(fname=target_fname, varname=varname)
+
+    def byte2s(b):
+        return "0x{},".format(hexlify(b).decode('utf-8'))
+
+    with open(local_fname, 'rb') as f, \
+         mmap.mmap(f.fileno(), 0, access=mmap.ACCESS_READ) as bfile:  # noqa: E127
+        yield from map(lambda x: x[1],
+            itertools.chain.from_iterable(     # noqa: E128
+                map(lambda l: itertools.chain(l[1], [(0, "\n")]),
+                    itertools.groupby(
+                        enumerate(map(byte2s, bfile)),
+                        lambda x: x[0]//16)
+                )  # noqa: E127
+            )
+        )
```

Comment:
wooha "the python foo is strong in this one" Not sure if this is more "nice and efficient" than it is "ugly and hard to understand/maintain"^^ I'm also wondering why you are mixing `yield from` and `itertools.chain`. I'm certain this can be written in a cleaner way, which might also help with getting rid of all the `#noqa ...` markers ;)